Auth use signalmap

jrajahalme opened the pull request:
Agent implementation of `pkg/signalmap` is generalized from a single target (conntrack GC) to allow for multiple targets. This is then used for passing policy notifications for required authentication from datapath to the agent instead of using the monitor perf ring buffer. This should reduce CPU overhead when monitoring is otherwise not used.

jrajahalme committed:
daemon: Remove eventsmap and signalmap from map status

Maximum nunber of entries in both eventsmap and signalmap is the number
of possible CPUs, which never changes, is a low number, and is not adding
any real information to the status report. Remove them to reduce direct
dependencies to these maps.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

christarazi commented on pkg/datapath/linux/node.go:
```diff
@@ -73,7 +73,8 @@ type linuxNodeHandler struct {
 
-	ipsecMetricCollector prometheus.Collector
+	ipsecMetricCollector  prometheus.Collector
+	ipsecMetricRegistered bool
```
Might be simpler to just use a `sync.Once` and avoid changing the flow by adding cases for the new boolean

jrajahalme replied:
Did you consider the dynamic of node configuration changing where ipsec is disabled/enabled? The straightforward solution with sync.Once would be:
- define a new helper that registers the metric only once, call that in the places where we register it now
- never unregister the metric

christarazi replied:
I don't think we support changing encryption on or off at runtime and so it requires a restart. So, this code will run again upon restart anyway.

jrajahalme replied:
So the code removing ipsec metric registration is dead code in the sense that it is only ever run on startup and if it is run, the metric was never registered (for the current invocation of the agent) anyway)?

jrajahalme replied:
Now that I reflect on this, I think I only hit the double registration issue in the integration tests due to other errors I had in the hive/cell changes. So maybe I can drop this commit from this PR? But it was confusing debugging those other issues when I saw the failures due to double registration of the ipsec metric!

jrajahalme replied:
@christarazi moved this to https://github.com/cilium/cilium/pull/25335

gandro reviewed (approved):
Looks good for my codeowners.

mhofstetter commented on daemon/cmd/daemon_main.go:
```diff
@@ -1590,6 +1591,7 @@ type daemonParams struct {
 	CertManager          certificatemanager.CertificateManager
 	SecretManager        certificatemanager.SecretManager
+	SignalManager        signal.SignalManager
 	AuthManager          auth.Manager
```
`AuthManager` is no longer used as parameter for the daemon -> remove

jrajahalme replied:
Removed, thanks!

mhofstetter commented on pkg/auth/cell.go:
```diff
@@ -39,46 +38,46 @@ var Cell = cell.Module(
+// Manager is an empty interface that is used by the daemon to pull in this module.
+type Manager interface{}
 
 func newManager(params authManagerParams) (Manager, error) {
```
with the removal of the AuthManager dependency in the daemon params the auth manager is no longer referenced by any other cell and therefore doesn't get initialized. 

Cell should be refactored to use `cell.Invoke(newManager)` instead of `cell.Provide(newManager),` -> auth manager no longer needs to be provided - it only needs to be initialized explicitly

jrajahalme replied:
Done, thanks!

mhofstetter commented on pkg/auth/cell.go:
```diff
@@ -39,46 +38,46 @@ var Cell = cell.Module(
-type Manager interface {
-	consumer.MonitorConsumer
-}
+// Manager is an empty interface that is used by the daemon to pull in this module.
```
interface can be deleted

jrajahalme replied:
Done.

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -37,30 +48,22 @@ type authHandler interface {
 }
 
-func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenticator, ipCache ipCache) (*authManager, error) {
+func newAuthManager(sm signal.SignalRegistrar, authHandlers []authHandler, dpAuthenticator datapathAuthenticator, ipCache ipCache, queueSize int) (*authManager, error) {
```
i would pass the pre-initialized signalChannel into the constructor and register the channel in the `cell`. this way these two aspects are separated.

jrajahalme replied:
Could you elaborate on this? What are the two aspects that need to be separated here and why? You mean registering the signal Channel to the signal package before calling into this constructor? Would I need to wrap the channel into a struct or can hive track a channel type just like a struct or an interface?

mhofstetter replied:
yes, i mean the registering the channel to the signal package before calling the constructor. IMO it would be already enough to have this part in the `newManager` function of the cell itself and then callilng the constructor only with the channel.

jrajahalme replied:
Done, thanks for the suggestion. This allowed the removal of the `SignalRegistrar` interface, as now the unit tests supply a channel directly.

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -72,123 +75,111 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
-	if err != nil {
-		return fmt.Errorf("failed to authenticate with auth type %s: %w", authType, err)
-	}
+func (a *authManager) Authenticate(key AuthKey) {
```
method `Authenticate` doesn't need to be exported -> find a better name to differentiate between the two clashing `authenticate` methods :thinking:

jrajahalme replied:
Moved this functionality to the calling site, which also makes pairing of `markPendingAuth()` and the corresponding `clearPendingAuth()` more explicit.

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -72,123 +75,111 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
-	}, nil
-}
+	if key.RemoteNodeID == 0 {
+		return fmt.Errorf("Auth required with local node, skipping")
```
as far as i understood the CfP, authentication should also be performed for the local node?

mhofstetter replied:
btw: it looks like this change breaks communication between pods on the same node if auth is enabled.

a local test ends up with having this message in the log without the call actually being passed through (timeout)

mhofstetter replied:
@jrajahalme works now as expected. thanks for fixing!

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -72,123 +75,111 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
-	}, nil
-}
+	if key.RemoteNodeID == 0 {
+		return fmt.Errorf("Auth required with local node, skipping")
```
nits:
* Error string should not be capitalized
* use `errors.New` instead `fmt.Errorf` if formatting isn't used.

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -72,123 +75,111 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
-		return hostIP
+	remoteNodeIP := a.ipCache.GetNodeIP(key.RemoteNodeID)
+	if remoteNodeIP == "" {
+		return fmt.Errorf("Remote node IP not available for node ID %d", key.RemoteNodeID)
```
nit: Error string should not be capitalized

mhofstetter commented on pkg/maps/cells.go:
```diff
@@ -21,6 +23,12 @@ var Cell = cell.Module(
 	configmap.Cell,
 
+	// Receives datapath signals for GC fill-up events
+	gc.Cell,
```
AFAIK this maps cell should only reference map cells. and not their inner cells (like GC in this case)

-> even though your introduced CT Map cell only contains the GC aspects for the time being, i would move the `cell.go` from package `ctmap/gc` to `gc`. In a later step the other aspects of the CT map can be addressed there as well.

jrajahalme replied:
You mean add an intermediary `ctmap/cell.go` so as not refer to the `gc.Cell` directly here? Or something else?

mhofstetter replied:
yes, or just move the `cell.go` completely. IMO the goal should be to have only the reference `ctmap.Cell` in this file (with an adapted description). GC is only a specific aspect of this map - even though it's the first aspect that got modularized! :+1:

jrajahalme replied:
Unfortunately this would create an import cycle, as `ctmap/gc` by definition needs to import `ctmap`. This means that a `ctmap/cell.go` can not import `ctmap/gc`.

I added a note in `maps/cell.go` to this effect.

mhofstetter replied:
might be worth to move the `gc.go` into the same package.

jrajahalme replied:
IMO that could be left for a separate exercise.

mhofstetter replied:
fine for me :+1:

mhofstetter commented on pkg/auth/manager.go:
```diff
@@ -72,123 +75,111 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
+	// starts later during the OnStart hook of the SignalManager
+	err := signal.RegisterChannel(sm, signalChannel, signal.SignalAuthRequired)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to set up signal channel for datapath authentication required events: %w", err)
```
nit: Error string should not be capitalized

jrajahalme replied:
Fixed, thanks!

mhofstetter commented on pkg/auth/cell.go:
```diff
@@ -39,46 +38,46 @@ var Cell = cell.Module(
 
 func (r config) Flags(flags *pflag.FlagSet) {
-	flags.Int("mesh-auth-monitor-queue-size", r.MeshAuthMonitorQueueSize, "Queue size for the auth monitor")
+	flags.Int("mesh-auth-queue-size", r.MeshAuthQueueSize, "Queue size for the auth manager")
```
should we add the actual purpose of "authentication" to the flag description? or maybe even mention the signalmap. currently it reads as THE queue of the auth manager. 

(we will have some more channels in the near future (GC, ..).)

jrajahalme replied:
Maybe we do not need to expose all of them separately to the user? For example, use a single queue-size parameter to control the set up of multiple queues/channels?

I'll leave this be for now, we can fine tune this when these additional queues are introduced.

mhofstetter replied:
agree :ok_hand: marking as resolved

mhofstetter reviewed (changes requested):
great improvement :rocket:  left some input from my side on the userspace code.

mhofstetter commented on pkg/signal/signal.go:
```diff
@@ -6,180 +6,286 @@ package signal
+
+	go func() {
+		log.Info("Datapath signal listener running")
+	For:
```
i would rather replace the inner switch with an if-elseif-else statement than going with the label break :)

jrajahalme replied:
Done :-)

mhofstetter commented on pkg/signal/signal.go:
```diff
@@ -6,180 +6,286 @@ package signal
+	return nil
+}
+
+// open the signalmap and start signal listener. Called after all the handlers have registered.
```
the signal map itself should already be open -> maybe change the description and/or change method name

jrajahalme replied:
Right, renamed these to `start()`/`stop()` and adjusted comments accordingly.

mhofstetter commented on pkg/signal/signal.go:
nit: contains some more capitalized error strings

jrajahalme replied:
Fixed, thanks for noticing :-)

mhofstetter commented on daemon/cmd/daemon_main.go:
```diff
@@ -1590,6 +1591,7 @@ type daemonParams struct {
 	EndpointManager      endpointmanager.EndpointManager
 	CertManager          certificatemanager.CertificateManager
 	SecretManager        certificatemanager.SecretManager
+	SignalManager        signal.SignalManager
```
`SignalManager` isn't used. it should be enough to have the signal cell itself registered in the daemon cell and have an explicit `invoke` in the cell itself (which both should be already the case).

or do we have some special initialization order requirements that i'm missing?

jrajahalme replied:
Not that I know of, removed this :-)

mhofstetter commented on pkg/signal/signal.go:
```diff
@@ -6,180 +6,286 @@ package signal
-	if signal >= SignalChanInvalid {
-		return fmt.Errorf("Signal number not supported: %d", signal)
+// RegisterChannel registers a go channel for the given signals.
+// 'sm' is not a receiver so that this can be a generic function.
```
an alternative approach would be to only provide a static wrapper which creates a `SignalHandler` for the given channel. without completely wrapping the `RegisterChannel` method.

```
// WriteToChannel creates a SignalHandler which writes the signal to the given channel.
func WriteToChannel[T fmt.Stringer](ch chan<- T) SignalHandler {
	return func(reader io.Reader) (string, error) {
		if reader == nil {
			close(ch)
			return "", io.EOF
		}
		var data T
		if err := binary.Read(reader, byteorder.Native, &data); err != nil {
			return "", err
		}
		select {
		case ch <- data:
		default:
			return "", ErrFullChannel
		}
		return data.String(), nil
	}
}
```

usage

```
if err := sm.RegisterHandler(signal.WriteToChannel(signalChannel), signal.SignalAuthRequired); err != nil {
	...
}
```

jrajahalme replied:
I like how this removes the need to pass the receiver as an explicit argument, thanks for the suggestion!

tommyp1ckles commented on pkg/datapath/linux/node_ids.go:
```diff
@@ -65,6 +65,14 @@ func (n *linuxNodeHandler) AllocateNodeID(nodeIP net.IP) uint16 {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	nodeIP, _ := n.nodeIPsByIDs[nodeID]
```
nit: this could just be `return n.nodeIPsByIDs[nodeID]`

jrajahalme replied:
Fixed, thanks!

mhofstetter commented on pkg/auth/cell.go:
```diff
@@ -39,46 +39,54 @@ var Cell = cell.Module(
 
 	Lifecycle             hive.Lifecycle
-	Config                config
+	SignalChannel         <-chan AuthKey
```
wasn't aware that this works with channels. might be a little bit problematic if someone else is providing a channel with the same type. but would result in an error - and therefore fine for the moment :+1:

jrajahalme committed:
signal: Use signalmap for auth required

Use signalmap instead of monitor events to notify userspace of missing
authentication.

Execute authentications concurrently and check for pending or completed
authentications before starting a new one with the same key.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
node: Add GetNodeIP() to NodeHandler

Add GetNodeIP(nodeID uint16) that returns the node IP for which the given
nodeID was allocated for. This will be used for authentication handshake
in later commits.

Check for 0 (== local node) explicitly and return local node address from
pkg/node in that case, as local node IPs are not in the maps.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter reviewed (approved):
LGTM :tada: thanks for addressing my comments

tommyp1ckles commented on pkg/auth/manager.go:
```diff
@@ -73,122 +75,96 @@ func newAuthManager(authHandlers []authHandler, dpAuthenticator datapathAuthenti
+					// updated the authmap since the datapath issued the auth
+					// required signal.
+					if a.datapathAuthenticator.checkAuthenticated(key) {
+						log.Debugf("auth: Already authenticated, skipped authentication for key %v", key)
```
should we be returning here?

jrajahalme replied:
Yes, messed this up inlining the former `Authenticate()` function here! Thanks for noticing :-)

jrajahalme committed:
signal: Generalize for multiple targets

Move definition of the signal data to the signal handlers to limit go
dependencies. Signal delivery channel is now buffered so that signal
manager will not get blocked by any one of the signal handlers.

Pause/Unpause is generalized so that the perf events are paused only
when all the signals have been paused. Current implementation uses bit
mask (64 bits) which limits the max number of signals to 64.

Signal registration is to be done only from init or cell OnStart
functions, so that no locking is needed for managing signal
handlers. Mute/Unmute calls can happen concurrently.

Manage dependencies via Hive/Cell.

Add a fake signalmap so that controlplane tests may depend on it.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

pippolo84 commented on pkg/maps/ctmap/gc/cell.go:
```diff
@@ -0,0 +1,59 @@
+// the SignalNatFillUp and SignalCTFillUp signals.
+type SignalData uint32
+
+const (
+	// SignalProtoV4 denotes IPv4 protocol
+	SignalProtoV4 SignalData = iota
+	// SignalProtoV6 denotes IPv6 protocol
+	SignalProtoV6
+	SignalProtoMax
+)
+
+var signalProto = [SignalProtoMax]string{
+	SignalProtoV4: "ipv4",
+	SignalProtoV6: "ipv6",
+}
```
Nit: we can get rid of the `SignalProtoMax` constant by letting the compiler calculate the size of the `signalProto` array with

```
var signalProto = [...]string{
	SignalProtoV4: "ipv4",
	SignalProtoV6: "ipv6",
}
```

pippolo84 commented on pkg/signal/signal.go:
```diff
@@ -6,180 +6,299 @@ package signal
+func ChannelHandler[T fmt.Stringer](ch chan<- T) SignalHandler {
+	closed := false
+	return func(reader io.Reader) (string, error) {
+		if ch == nil {
+			return "", ErrNilChannel
+		}
```
I suggest to move this out of the returned func, since `ch` is not gonna change after the call to `ChannelHandler`. Doing that we can also avoid generating a func that always returns `"", ErrNilChannel`.

jrajahalme commented on pkg/maps/signalmap/cell.go:
```diff
@@ -0,0 +1,59 @@
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "signalmap")
+)
```
Will drop this `log` in favour of `logrus.FieldLogger` provided by Hive in a forthcoming update.

pippolo84 commented on pkg/auth/cell.go:
```diff
@@ -39,46 +39,54 @@ var Cell = cell.Module(
-
-func newManager(params authManagerParams) (Manager, error) {
-	mgr, err := newAuthManager(params.AuthHandlers, params.DatapathAuthenticator, params.IPCache)
+func newManager(params authManagerParams) error {
```
Nit: we usually call functions to be invoked as `registerX`, leaving `newX` for the ones that provide API to the hive.

pippolo84 reviewed (approved):
Overall LGTM, just some non-blocking nits. :100:

jibi reviewed (approved):
ipsec changes lgtm :+1:
